irobotclient/request_handler.py

```diff
@@ -0,0 +1,161 @@
+"""
+Copyright (c) 2017 Genome Research Ltd.
+
+This program is free software: you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation, either version 3 of the License, or (at your
+option) any later version.
+
+This program is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import requests
+import time
+import errno
+
+from datetime import datetime
+from collections import namedtuple
+
+from irobotclient.custom_exceptions import IrobotClientException
+
+"""
+Limit on how many times the same request is sent, including any alterations such as changes to the header.
+In theory a single request will take two attempts; one for the HEAD request to make sure the data exists and
+a second to GET the data.
+
+However, a request may generate a response to wait for the data to be fetched and try again later or to attempt
+a different authorisation method.
+
+Either way, the limit defined below helps prevent stack overflow and unnecessary requests when it is clear a request
+is not going to work.
+"""
+REQUEST_LIMIT = 10
+
+"""
+If a 202 response returns with not iRobot-ETA header then a default delay (in seconds) will be set.
+"""
+DEFAULT_REQUEST_DELAY = 600
+
+"""
+The iRobot response status codes are declared below along with an associated standard error number if applicable.
+"""
+ResponseStruct = namedtuple("ResponseStruct", "status_code, errno")
+
+RESPONSES = {
+    'SUCCESS': ResponseStruct(status_code=200, errno=None),
+    'FETCHING_DATA': ResponseStruct(status_code=202, errno=None),
+    'RANGED_DATA': ResponseStruct(status_code=206, errno=None),
+    'CLIENT_MATCHED': ResponseStruct(status_code=304, errno=None),
+    'AUTH_FAIL': ResponseStruct(status_code=401, errno=errno.ECONNREFUSED),
+    'DENIED_IRODS': ResponseStruct(status_code=403, errno=errno.EACCES),
+    'NOT_FOUND': ResponseStruct(status_code=404, errno=errno.ENODATA),
+    'INVALID_REQUEST_METHOD': ResponseStruct(status_code=405, errno=errno.EPROTO),
+    'INVALID_MEDIA_REQUESTED': ResponseStruct(status_code=406, errno=errno.EINVAL),
+    'INVALID_RANGE': ResponseStruct(status_code=416, errno=errno.ERANGE),
+    'TIMEOUT': ResponseStruct(status_code=504, errno=errno.ETIMEDOUT),
+    'PRECACHE_FULL': ResponseStruct(status_code=507, errno=errno.ENOMEM)
+}
+
+
+class Requester:
+    def __init__(self, requested_url: str, headers: dict):
+        """
+
+        :param requested_url:
+        :param headers:
+        """
+
+        self._request = requests.Request(url=requested_url, headers=headers)
+        self._request_delay = 0
+
+    def get_data(self) -> requests.Response:
```

Comment:
A custom model would just be a wrapper exposing the functionality you need. You can then inject `requests` and massage its API to match yours. If one day you decided that `requests` sucked and you wanted to instead use `super-requests`, you would just have to write a new class, abstracted from your model, injecting `super-requests`, and the rest of your code would stay the same.

IMO, this is a "nice to have", but not a priority for core dependencies. For example, I don't usually go to this level of abstraction, but that's not to go against Colin's suggestion: It is a generally good idea.